library/sonar_plugin.py

```diff
@@ -0,0 +1,429 @@
+    response = requests.get(url, auth=HTTPBasicAuth(module.params['username'],
+                                                    module.params['password']))
+    json_obj = response.json()
+
+    for plugin in json_obj['plugins']:
+        if plugin['name'] == module.params['name']:
+            version_installed = plugin['version']
+
+    comparison = compare_plugins_version(module.params['version'],
+                                         version_installed)
+
+    if comparison == 'Equal':
+        module.exit_json(changed=False, stdout='FAILED: Version already installed')
+    elif comparison == 'Lower':
+        is_available = True
+    else:
+        url = sonar_url + updates_list_api
+        response = requests.get(url, auth=HTTPBasicAuth(module.params['username'], module.params['password']))
+        json_obj = response.json()
+        for plugin in json_obj['plugins']:
+            if plugin['name'] == module.params['name']:
+                for update in plugin['updates']:
+                    if compare_plugins_version(module.params['version'], update['release']['version']) == 'Equal':
+                        if not update['requires'] and update['status'] == 'COMPATIBLE':
+                            is_available = True
+
+    return is_available
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            name=dict(type='str'),
+            state=dict(type='str', choices=['installed', 'removed', 'latest']),
+            version=dict(type='str'),
+            custom_url=dict(type='str'),
+            hostname=dict(default='localhost', type='str'),
+            username=dict(default='admin', type='str'),
+            password=dict(default='admin', type='str', no_log=True),
+            pending_dir=dict(default='/usr/local/sonar/extensions/downloads/', type='str'),
+            sonar_port=dict(default='9000', type='int')
+        ),
+        mutually_exclusive=[['custom_url', 'state']]
+    )
+
+    available_list_api = '/api/plugins/available'
+    installed_list_api = '/api/plugins/installed'
+    updates_list_api = '/api/plugins/updates'
+
+    install_api = '/api/plugins/install'
+    remove_api = '/api/plugins/uninstall'
+    update_api = '/api/plugins/update'
+
+    sonarsource_plugins_repo = 'https://binaries.sonarsource.com/Distribution/'
+
+    state = module.params['state']
+    sonar_url = 'http://' + module.params['hostname'] + ':' + str(module.params['sonar_port'])
+    plugin_data = {}
+
+    if module.params['custom_url']:
+        status_pending = 'custom'
+        key = download_custom_plugin(module.params['custom_url'], module.params['pending_dir'], status_pending)
+        if key != 'not-found':
+            if is_plugin_pending(key, status_pending, module):
+                msg = 'SUCCESS: The plugin with key={} is pending'.format(key)
+                module.exit_json(changed=True, stdout=msg)
+        else:
+            module.exit_json(changed=True,
```

Comment:
is it a success?

Reply:
yes. we talked about it and I added some pre-checks